### PR TITLE
[WIP] quic

### PR DESCRIPTION
**Pull request by bagder**
This is **early** (and incomplete) work on quic support. Pushed as a PR only to make sure the work so far hasn't broken the build in unrelated parts (by running it through the CI tests).

The first steps of the plan;
- [x] - make all QUIC code conditional on the detection of ngtcp2 at build time
- [x] - support ngtcp2 in configure
- [ ] - support ngtcp2 in cmake
- [x] - show "HTTP3" as a feature in `--version` output
- [x] - show ngtcp2 in the version string for third party libs
- [ ] - figure out how to show the *correct* ngtcp2 version number in that string
- [x] - support `CURLOPT_H3` and the `CURLH3_DIRECT` flag
- [x] - support `--http3-direct` in the command line tool
- [ ] - actually connect using QUIC when `CURLH3_DIRECT` is set
- [ ] - disconnect QUIC

Alt-Svc support (for HTTP/3 and other versions) is being worked on separately by @xquery.

There's no general HTTP/3 support for ngtcp2 available yet.

[Wiki page for QUIC implementation thoughts](https://github.com/curl/curl/wiki/QUIC-implementation)

### CODE REVIEW

**Comment by gvanem** [comment]
Ops, sorry for the edit (undid it). But I'll try to build with this.

File: lib/quic.c [hunk 1]
```diff
@@ -0,0 +1,983 @@
...
+  uint8_t pn[64];
+  ssize_t pnlen;
+  struct Context *crypto_ctx = &conn->quic.crypto_ctx;
+
+  switch(name) {
+  case SSL_KEY_CLIENT_EARLY_TRAFFIC:
+  case SSL_KEY_CLIENT_HANDSHAKE_TRAFFIC:
+  case SSL_KEY_CLIENT_APPLICATION_TRAFFIC:
+  case SSL_KEY_SERVER_HANDSHAKE_TRAFFIC:
+  case SSL_KEY_SERVER_APPLICATION_TRAFFIC:
```
Comment (gvanem) [inline lib/quic.c]: Error here also:
```
quic.c(307,8):  error: use of undeclared identifier 'SSL_KEY_CLIENT_EARLY_TRAFFIC'
  case SSL_KEY_CLIENT_EARLY_TRAFFIC:
       ^
quic.c(308,8):  error: use of undeclared identifier 'SSL_KEY_CLIENT_HANDSHAKE_TRAFFIC'
  case SSL_KEY_CLIENT_HANDSHAKE_TRAFFIC:
       ^
quic.c(309,8):  error: use of undeclared identifier 'SSL_KEY_CLIENT_APPLICATION_TRAFFIC'
  case SSL_KEY_CLIENT_APPLICATION_TRAFFIC:
       ^
quic.c(310,8):  error: use of undeclared identifier 'SSL_KEY_SERVER_HANDSHAKE_TRAFFIC'
  case SSL_KEY_SERVER_HANDSHAKE_TRAFFIC:
       ^
```

File: lib/quic.c [hunk 2]
```diff
@@ -0,0 +1,983 @@
...
+          ERR_error_string(ERR_get_error(), NULL));
+    return NULL;
+  }
+
+  if(SSL_CTX_set1_groups_list(ssl_ctx, QUIC_GROUPS) != 1) {
+    failf(data, "SSL_CTX_set1_groups_list failed");
+    return NULL;
+  }
+
+  SSL_CTX_set_mode(ssl_ctx, SSL_MODE_QUIC_HACK);
```
Comment (gvanem) [inline lib/quic.c]: I get this error:
```
quic.c(269,29):  error: use of undeclared identifier 'SSL_MODE_QUIC_HACK'
```

File: lib/quic.c (hunk 1 above)
Comment (bagder) [reply lib/quic.c]: Those are symbols that are added in ngtcp2's custom openssl version. Did you build and install that and made sure to point out that build when you build this curl? I mention this in the docs/HTTP3.md file but I figure it is easily missed!

Comment (gvanem) [reply lib/quic.c]: >  Did you build and install that ..

No, I missed that. 

BTW. it seems ngtcp2 is not yet quite Windows-ready. Includes `<unistd.h>`, but not `<winsock2.h>` etc.

Comment (gvanem) [reply lib/quic.c]: I now d/l and built with this `https://github.com/tatsuhiro-t/openssl/`. The `SSL_MODE_QUIC_HACK` etc. is nowhere to be found there.  Are you sure your WIP is up-to-date against Tatsuhiro's OpenSSL?

Comment (bagder) [reply lib/quic.c]: That define is used by ngtcp2's own client.cc example:

https://github.com/ngtcp2/ngtcp2/blob/c0cd76426965fd7162132523a9dd04e6a2f9c200/examples/client.cc#L1941

... even in their draft-15 branch which I believe is where they're working mostly right now.

I'm using the quic-draft-14 branch of @tatsuhiro-t's openssl (as the ngtcp2 README says so) and that defines `SSL_MODE_QUIC_HACK` here:

https://github.com/tatsuhiro-t/openssl/blob/bfd4950424bfc1508f1e70cc0ca9de772ed38b11/include/openssl/ssl.h#L497

> it seems ngtcp2 is not yet quite Windows-ready

I'm sure they'll appreciate a pull-request that remedies that! =)

**Comment by gvanem** [comment]
Finally got a version to try. But a `curl --http3-direct -v https://www.google.com` just creates a lot of `SOCK_DGRAM` sockets and then gives up with `curl: (2) Failed initialization`. But I know next to nothing about QUIC.

But an issue when building with `-DUSE_SCHANNEL` and this *Tatsuhiro* OpenSSL is the clash of  `<wincrypt.h>` names (just like in BoringSSL). I did this to fix:
```diff
--- a/lib/vtls/schannel.h 2018-11-02 23:47:24
+++ b/vtls/schannel.h 2018-12-12 16:13:40
@@ -45,10 +45,12 @@
  * BoringSSL's <openssl/x509.h>: So just undefine those defines here
  * (and only here).
  */
-#if defined(HAVE_BORINGSSL) || defined(OPENSSL_IS_BORINGSSL)
+#if defined(HAVE_BORINGSSL) || defined(OPENSSL_IS_BORINGSSL) || defined(OPENSSL_IS_Tatsuhiro)
 # undef X509_NAME
 # undef X509_CERT_PAIR
 # undef X509_EXTENSIONS
+# undef OCSP_REQUEST
+# undef OCSP_RESPONSE
 #endif
```

Anyway, this was the OpenSSL I had to use and build with:

```
 git clone --recursive --branch quic-draft-15 https://github.com/tatsuhiro-t/openssl.git .
```
Any hope this will crawl it's way into the official OpenSSL? I hate to have several OpenSSL beasts on my disk.

File: lib/quic.c [hunk 3]
```diff
@@ -0,0 +1,983 @@
...
+  return CURLE_OK;
+}
+
+/*
+ * Store ngtp2 version info in this buffer, Prefix with a space.  Return total
+ * length written.
+ */
+int Curl_quic_ver(char *p, size_t len)
+{
+  return msnprintf(p, len, " ngtc2/blabla");
```
[the remark this answers is not in the document]
Comment (bagder) [reply lib/quic.c]: I expect the library will offer a function that returns the version number, but it doesn't yet so...

**Comment by bagder** [comment]
> Finally got a version to try. But a curl --http3-direct -v https://www.google.com just creates a lot of SOCK_DGRAM sockets and then gives up with curl: (2) Failed initialization. But I know next to nothing about QUIC.

Right, I've written a lot of glue code but I've not yet made it work.  It's really complicated for me to work out where any problem might lie since I too lack the in-depth knowledge of QUIC, ngtcp2 and TLS that require. My work so far has been to see what the client.cc code does and then converting that into C code for curl.

It also just fires off a single function call to ngtcp2 so it doesn't even try to do the right thing yet.

> But an issue when building with -DUSE_SCHANNEL and this

I've decided to start off with building with OpenSSL only and then only the Tatsuhiro OpenSSL version. I figure we'll deal with multi-SSL and other-SSL issues later.

> Any hope this will crawl it's way into the official OpenSSL?

Let's hope so. Otherwise we'll have a really painful situation.

File: lib/quic.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (gvanem) [reply lib/quic.c]: I was referring to the missing `p`.

**Comment by tatsuhiro-t** [comment]
I don't think google is deploying IETF QUIC which is still in draft stage, and changing rapidly and drastically.
Here is the existing implementations and its endpoint: https://github.com/quicwg/base-drafts/wiki/Implementations
Note that the information might  not be up to date.

For OpenSSL stuff, someone has to ask OpenSSL dev to implement required features perhaps after QUIC draft is sent to IESG.  I heard that BoringSSL has an implementation for these features, let's hope that they can be added to OpenSSL.  I don't know about LibreSSL.  Supporting similar but slightly different TLS backends is just a pain.

**Comment by gvanem** [comment]
> It also just fires off a single function call to ngtcp2 so it doesn't even try to do the right thing yet.

Same here. Using Tatsuhiro's test server:
```
curl.exe -v4 --http3-direct https://nghttp2.org:4433
```
I get an error in `setup_initial_crypto_context()` and `Curl_qc_derive_initial_secret()` (line 69 of `quic.c`).
And nothing gets sent.

No idea what the `hkdf_extract()` is all about.

File: docs/HTTP3.md [hunk 4]
```diff
@@ -0,0 +1,46 @@
+# HTTP3 (and QUIC)
+
+## Resources
+
+[HTTP/3 Explained](https://daniel.haxx.se/http3-explained/) - the online free
+book describing the protocols involved.
+
+[QUIC implementation](https://github.com/curl/curl/wiki/QUIC-implementation) -
+the wiki page describing the plan for how to support QUIC and HTTP/3 in cul
```
Comment (danielgustafsson) [inline docs/HTTP3.md]: s/cul/curl/

File: docs/cmdline-opts/http3-direct.d [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+Tags: Versions
+Protocols: HTTP
+Added: 7.66.0
+Mutexed: http1.1 http1.0 http2 http2-prior-knowledge
+Requires: HTTP/3
+Help: Use HTTP v3
+---
+
+Tells curl to use HTTP version 3 directly to the host and port number used in
+the URL. A normal HTTP/3 transation will be done to a host and then get
```
Comment (danielgustafsson) [inline docs/cmdline-opts/http3-direct.d]: s/transation/transaction/

File: lib/quic.c [hunk 6]
```diff
@@ -0,0 +1,983 @@
...
+  quic_callbacks(&qs->callbacks);
+
+#ifdef NGTCP2_PROTO_VER_D15
+#define QUICVER NGTCP2_PROTO_VER_D15
+#else
+#define QUICVER NGTCP2_PROTO_VER_D14
+#endif
+  /* ngtcp2 master branch uses version NGTCP2_PROTO_VER_D14 */
+  rc = ngtcp2_conn_client_new(&qs->conn, &qs->dcid, &qs->scid,
+                              QUICVER, &qs->callbacks, &qs->settings, conn);
```
Comment (tatsuhiro-t) [inline lib/quic.c]: It looks like qs->dcid and qs->scid are not initialized.
They must be some random bytes.

File: lib/connect.c [hunk 7]
```diff
@@ -1122,6 +1123,13 @@ static CURLcode singleipconnect(struct connectdata *conn,
     if(-1 == rc)
       error = SOCKERRNO;
   }
+#ifdef USE_NGTCP2
+  else if(!isconnected && (conn->transport == TRNSPRT_QUIC)) {
+    result = Curl_quic_connect(conn, sockfd, &addr.sa_addr, addr.addrlen);
+    if(result)
+      return result;
+  }
```
Comment (tatsuhiro-t) [inline lib/connect.c]: It looks like rc is still -1, which makes this function fail.

**Comment by ghedo** [comment]
Shameless plug, but I hacked together QUIC support for curl using my own [quiche](https://github.com/ghedo/quiche) library. I based my changes on the branch from this PR and replaced the ngtcp2-specific bits and a few other things.

The end result (but still incompete and WIP) can be seen at https://github.com/ghedo/curl/tree/quiche (build instructions in the commit message)

I've got to the point where I can do a single HTTP/1.1 request and get a response from a QUIC server:
```
 % src/curl --http3-direct https://cloudflare-quic.com/   
       ____    
      / __ \   
     / / / /   
    / /_/ /    
    \___\_\    
       __  __  
      / / / /  
     / / / /   
    / /_/ /    
    \____/     
        ____   
       /  _/   
       / /     
     _/ /      
    /___/      
      ______   
     / ____/   
    / /        
   / /___      
   \____/
```

Compared to ngtcp2, quiche requires _a lot_ less glue code, but on the other hand it depends on BoringSSL (you can actually use latest BoringSSL master, so no additional patches are required, but still), so building is a bit more complex since right now curl also needs to be built against bssl to avoid conflicts with OpenSSL.

**Comment by bagder** [comment]
Awesome @ghedo, I will jump in and try this out soonish.

I would like to make sure that we unify our efforts and make sure that we can both move forward and still have the same take on quic in general within curl and make sure that we can build/use either library according to configure options.

**Comment by ghedo** [comment]
@bagder Yes, I agree. This was mostly an exercise to check if the quiche API can be integrated into existing code, and it takes some short-cuts so I'm not convinced it's the correct way to go. Though I hope that having something that "works" (even if hacky) can make iterating over the curl internals changes easier.

**Comment by bagder** [comment]
I'll clean up my work a little bit and make sure we can build and use it with *either* ngtcp2 or quiche so that we can experiment better going forward. Gimme a day or two and then I'll present something that should be a little better foundation to build on. I think we'll benefit from that and I really like that we can already do HTTP requests with your branch.

**Comment by bagder** [comment]
@ghedo can you give my branch at https://github.com/curl/curl/tree/QUIC a glance and see if you can figure out how I botched quiche from working there as nicely as in your repo?

It is my attempt in making QUIC supported by functions that can be implemented by either of the backends with both ngtcp2 and (your) quiche code in the repo.

**Comment by ghedo** [comment]
The problem seems to be in `singleipconnect()`. In my changes I made the `connect()` syscall happen on the UDP socket as well, because otherwise I couldn't figure out where to get the peer address to do `sendto()` instead of just `send()` 

Basically you are missing the following change from my branch:

```diff
diff --git a/lib/connect.c b/lib/connect.c
index b11df23cb..8db5a0489 100644
--- a/lib/connect.c
+++ b/lib/connect.c
@@ -1076,7 +1076,7 @@ static CURLcode singleipconnect(struct connectdata *conn,
     Curl_expire(data, conn->timeoutms_per_addr, EXPIRE_DNS_PER_NAME);
 
   /* Connect TCP sockets */
-  if(!isconnected && (conn->transport == TRNSPRT_TCP)) {
+  if(!isconnected) {
     if(conn->bits.tcp_fastopen) {
 #if defined(CONNECT_DATA_IDEMPOTENT) /* Darwin */
 #  if defined(HAVE_BUILTIN_AVAILABLE)
@@ -1123,18 +1123,19 @@ static CURLcode singleipconnect(struct connectdata *conn,
     if(-1 == rc)
       error = SOCKERRNO;
   }
+  else {
+    *sockp = sockfd;
+    return CURLE_OK;
+  }
+
 #ifdef ENABLE_QUIC
-  else if(!isconnected && (conn->transport == TRNSPRT_QUIC)) {
+  if(!isconnected && (conn->transport == TRNSPRT_QUIC)) {
     result = Curl_quic_connect(conn, sockfd, &addr.sa_addr, addr.addrlen);
     if(result)
       return result;
     rc = 0; /* connect success */
   }
 #endif
-  else {
-    *sockp = sockfd;
-    return CURLE_OK;
-  }
 
   if(-1 == rc) {
     switch(error) {
```

This is one of the hacks I had to do to get this going. IIRC the `ip_addr` and `tempaddr` fields of `struct connectdata` weren't fully initialized for some reason, so I just made the socket be "connected".

**Comment by lilianmoraru** [comment]
@bagder Note that the instructions for compiling both [`BoringSSL`](https://github.com/curl/curl/compare/QUIC#diff-02fa1449f22ba0ba3f058066b30389a3R61) and [`quiche`](https://github.com/curl/curl/compare/QUIC#diff-02fa1449f22ba0ba3f058066b30389a3R71) suggest compiling without optimizations(if it was intentional, please ignore the next lines :smile: ).
For `BoringSSL`:
```
cmake -DCMAKE_POSITION_INDEPENDENT_CODE=on -DCMAKE_BUILD_TYPE=RelWithDebInfo .. # for an optimized build with debug info
cmake -DCMAKE_POSITION_INDEPENDENT_CODE=on -DCMAKE_BUILD_TYPE=Release .. # for an optimized build _without_ debug info
```

For `quiche`:
```
QUICHE_BSSL_PATH=$PWD/deps/boringssl cargo build --release # for an optimized build with debug info(`quiche` sets debug = true in `Cargo.toml`)
QUICHE_BSSL_PATH=$PWD/deps/boringssl RUSTFLAGS="-C debuginfo=0" cargo build --release # optimized + forced removal of debug info
```

**Comment by bagder** [comment]
Thanks a lot @ghedo, I adjusted that patch a little, amended my commits and forced-pushed my QUIC branch just now. Since PRs are tied so much to a specific branch, I'll close this and create a new one for that branch which now is a more suitable way forward.

**Pull request closed without merging**